Changelog v1.13 (#77)
meilisearch/engine-team: meilisearch-release-changelog/v1.13.0.md

curl \
  -X PATCH 'MEILISEARCH_URL/network/' \
  -H 'Content-Type: application/json'  \
  --data-binary '{
    "remotes": {
      "ms-0": {
        "url": "http://ms-1235.example.meilisearch.io",
        "searchApiKey": "Ecd1SDDi4pqdJD6qYLxD3y7VZAEb4d9j6LJgt4d6xas"
      },
      "ms-1": {
        "url": "http://ms-4242.example.meilisearch.io",
        "searchApiKey": "hrVu-OMcjPGElK7692K7bwriBoGyHXTMvB5NmZkMKqQ"
      }
    },
    "self": "ms-0"
  }'
```
Repeat this process with every instance in your network, then add documents to them. Do not send the same documents to different instances.
Finally, make a `/multi-search` query with the new `federationOptions.remote`:

```sh
curl \
  -X PATCH 'MEILISEARCH_URL/multi-search/' \
  -H 'Content-Type: application/json'  \
  --data-binary '{
    "federation": {},
    "queries": [
        {
            "q": "Batman returns dark",
            "indexUid": "movies",
            "federationOptions": {
                "remote": "ms-0"
            }
        },
        {
            "q": "Batman returns dark",
            "indexUid": "movies",
            "federationOptions": {
                "remote": "ms-1"
            }
        }
    ]

> }

Reviewer: ```suggestion
}'
```

  * Log more metrics around HTTP embedder requests and ANNs in arroy by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5288
  * Fix internal error when not correctly using the `documentTemplateMaxBytes` parameter by @dureuill in https://github.com/meilisearch/meilisearch/pull/5306
  * When a batch is deleted it no longer keeps some internal data in the DB @Kerollmops and @irevoire in https://github.com/meilisearch/meilisearch/pull/5272
  * Fix a rare stack overflow when using remote embedders by @Kerollmops with the help of @dureuill https://github.com/meilisearch/meilisearch/pull/5294 (from v1.12.8)
  * Fix Dotnet tests in sdks-tests.yml by @curquiza in https://github.com/meilisearch/meilisearch/pull/5291
  * Debug log the channel congestion by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5177
  * Improve unexpected panic message by @irevoire in https://github.com/meilisearch/meilisearch/pull/5300
  * Introduce a compaction subcommand in meilitool by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5235

❤️ Thanks again to our external contributors:

> - [Meilisearch](https://github.com/meilisearch/meilisearch): @takaebato, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay, @DerTimonius, @manojks1999

Reviewer 2: ```suggestion
- [Meilisearch](https://github.com/meilisearch/meilisearch): @takaebato, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay, @DerTimonius, @manojks1999, @ellnix, @Guikingone
```

I added some issue reporters

  * When a batch is deleted it no longer keeps some internal data in the DB @Kerollmops and @irevoire in https://github.com/meilisearch/meilisearch/pull/5272
  * Fix a rare stack overflow when using remote embedders by @Kerollmops with the help of @dureuill https://github.com/meilisearch/meilisearch/pull/5294 (from v1.12.8)
  * Fix Dotnet tests in sdks-tests.yml by @curquiza in https://github.com/meilisearch/meilisearch/pull/5291
  * Debug log the channel congestion by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5177
  * Improve unexpected panic message by @irevoire in https://github.com/meilisearch/meilisearch/pull/5300
  * Introduce a compaction subcommand in meilitool by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5235
  * Expose a route to get the file content associated with a task by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5326

❤️ Thanks again to our external contributors:
- [Meilisearch](https://github.com/meilisearch/meilisearch): @takaebato, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay, @DerTimonius, @manojks1999, @ellnix, @Guikingone, @migueltarga

> - [Charabia](https://github.com/meilisearch/charabia): <TBD: exhaustive list of external contributors of Charabia>

Reviewer 3: ```suggestion
```

### Breaking Changes before Stabilization

* Accepted Ollama URLs can only end with `/api/embed` and `/api/embeddings`.
* Modified error codes:
  - `invalid_embedder` has been split into `invalid_search_embedder` and `invalid_similar_embedder`. These codes are returned when the `embedder` parameter of a search (resp. similar) request refers to a non-existing embedder configuration or is not a string.
  - `invalid_hybrid_query` has been renamed to `invalid_search_hybrid_query`. It is returned when the `hybrid` parameter is invalid: contains unknown keys or is not either `null` or an object.

Done by @dureuill in https://github.com/meilisearch/meilisearch/pull/5232 & https://github.com/meilisearch/meilisearch/pull/5234

## Dumpless Upgrade: Ease upgrading to the next version without a dump

TBD by tamo

## Proxy search requests

> TBD by Louis

Reviewer 2: ````suggestion
You can now transparently federate search results from multiple Meilisearch instances.

1. Enable the `proxy_search` experimental feature.
2. Setup your network of multiple instances with a call to `PATCH /network`
```json
{
  "remotes": {
      "ms-0": {
        "url": "http://ms-1235.example.meilisearch.io",
        "searchApiKey": "Ecd1SDDi4pqdJD6qYLxD3y7VZAEb4d9j6LJgt4d6xas"
      },
      "ms-1": {
	      "url": "http://ms-4242.example.meilisearch.io",
	      "searchApiKey": "hrVu-OMcjPGElK7692K7bwriBoGyHXTMvB5NmZkMKqQ"
	    }
  },
  "self": "ms-0"
}
```
3. Create the same indexes with the same settings on all your Meilisearch instances.
4. Send your documents to your instances, making sure to never send the same documents to different Meilisearch instances.
5. Use the new `federationOptions.remote` search query parameter in your federated search on `POST /multi-search`
```json
{
    "federation": {
        "limit": 5
    },
    "queries": [
        {
            "q": "Batman returns dark",
            "indexUid": "movies",
            "federationOptions": {
                "remote": "ms-0"
            },
            "attributesToRetrieve": [
                "title"
            ]
        },
        {
            "q": "Batman returns dark",
            "indexUid": "movies",
            "federationOptions": {
                "remote": "ms-1"
            },
            "attributesToRetrieve": [
                "title"
            ]
        }
    ]
}
```
This yields the following results:
```json
{
  "hits": [
    {
      "title": "Batman: The Dark Knight Returns, Part 1",
      "_federation": {
        "indexUid": "mieli",
        "queriesPosition": 1,
        "weightedRankingScore": 0.939452124935996,
        "remote": "ms-1"
      }
    },
    {
      "title": "Batman: The Dark Knight Returns, Part 2",
      "_federation": {
        "indexUid": "mieli",
        "queriesPosition": 1,
        "weightedRankingScore": 0.939452124935996,
        "remote": "ms-1"
      }
    },
    {
      "title": "Batman Beyond: Return of the Joker",
      "_federation": {
        "indexUid": "mieli",
        "queriesPosition": 0,
        "weightedRankingScore": 0.8660714285714286,
        "remote": "ms-0"
      }
    },
    {
      "title": "Batman Returns",
      "_federation": {
        "indexUid": "mieli",
        "queriesPosition": 0,
        "weightedRankingScore": 0.6666666666666666,
        "remote": "ms-0"
      }
    },
    {
      "title": "Batman: Return of the Caped Crusaders",
      "_federation": {
        "indexUid": "mieli",
        "queriesPosition": 1,
        "weightedRankingScore": 0.5798059964726632,
        "remote": "ms-1"
      }
    }
  ],
  "processingTimeMs": 160,
  "limit": 5,
  "offset": 0,
  "estimatedTotalHits": 64,
  "remoteErrors": {}
}
```

Using this new feature, you can split a large index across multiple Meilisearch instances.
Find more information about this feature in the [public usage page](https://meilisearch.notion.site/API-usage-Remote-search-request-f64fae093abf409e9434c9b9c8fab6f3?pvs=74)

````

* Speed up the request to list the indexes by @irevoire in https://github.com/meilisearch/meilisearch/pull/5166
* Add support for `GITHUB_TOKEN` authentication in installation script by @Sherlouk in https://github.com/meilisearch/meilisearch/pull/5216

# Fixes 🐞

* Improve the error message when an attributes is not filterable by @jameshiew in https://github.com/meilisearch/meilisearch/pull/5135

# Misc

* Dependencies updates
  * Upgrade dependencies and fixes the `idna` severity issue by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5218
* CIs and tests
  * Split Meilisearch crate tests in separate file by @K-Kumar-01 in https://github.com/meilisearch/meilisearch/pull/5134
  * Split tests in separate file by @K-Kumar-01 in https://github.com/meilisearch/meilisearch/pull/5171
  * Remove obsolete test code by @K-Kumar-01 in https://github.com/meilisearch/meilisearch/pull/5173
  * Fix the flaky batches test by @irevoire in https://github.com/meilisearch/meilisearch/pull/5175
  * Split tests for option crate meilisearch in a separate test file by @K-Kumar-01 in https://github.com/meilisearch/meilisearch/pull/5174

>   * Remove hard coded task IDs to prevent flaky tests by @mhmoudr in https://github.com/meilisearch/meilisearch/pull/5182

Reviewer 3: ```suggestion
  * Remove hard coded task IDs to prevent flaky tests by @mhmoudr in https://github.com/meilisearch/meilisearch/pull/5182
  * Improve test performance of get_index.rs by @DerTimonius in https://github.com/meilisearch/meilisearch/pull/5210
```

* Misc
  * Guide people to create custom reports on the benchboard by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5029
  * Fix typo in a comment by @eltociear in https://github.com/meilisearch/meilisearch/pull/5184
  * Replace hardcoded string with constants by @Gnosnay in https://github.com/meilisearch/meilisearch/pull/5169
  * Refactor the index-scheduler by @irevoire in https://github.com/meilisearch/meilisearch/pull/5199
  * Refactor indexer by @dureuill in https://github.com/meilisearch/meilisearch/pull/5168
  * Auto-generate the OpenAPI spec by @irevoire in https://github.com/meilisearch/meilisearch/pull/4867 & https://github.com/meilisearch/meilisearch/pull/5231
  * Merge bitmaps by using `Extend::extend` by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5221

>   * Fix corrupted task queue errors on index creation by @irevoire in https://github.com/meilisearch/meilisearch/pull/5239

Reviewer 3: ```suggestion
  * Fix issue introduced by dumpless upgrade by @manojks1999 in https://github.com/meilisearch/meilisearch/pull/5284
```

❤️ Thanks again to our external contributors:

> - [Meilisearch](https://github.com/meilisearch/meilisearch): @dureuill, @Kerollmops, @takaebato, @irevoire, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay.

Reviewer 3: ```suggestion
- [Meilisearch](https://github.com/meilisearch/meilisearch): @takaebato, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay.
```

Author: Thank you 😂

# v1.13.0 release changelogs

<!-- The following line should ONLY be put PRE-release changelogs -->
> [!WARNING]
> Since this is a release candidate (RC), we do NOT recommend using it in a production environment. Is something not working as expected? We welcome [bug reports](https://github.com/meilisearch/meilisearch/issues/new/choose) and [feedback about new features](https://github.com/meilisearch/product/discussions).

> Meilisearch v1.13 introduces several significant improvements: our Vector Store feature is now fully stabilized for seamless hybrid search integration, dumpless version upgrades eliminate the need for reindexing during updates, granular filterable attribute settings allow for optimized indexing performance, and a new experimental feature enables distributed search queries across multiple Meilisearch instances to better balance network loads.

Reviewer 3: ```suggestion
Meilisearch v1.13 introduces several significant improvements including the stabilization of the Vector Store feature for seamless hybrid search integration! 🎉 This version also ease the Meilisearch upgrade by simplifying the process: you don't need dump for migrating anymore! Finally, we introduced more granular filterable attribute settings to optimize your indexing performance.
```

Reviewer 3: I suggest something in multiple sentences, easier to read, and more focus on the value for the users than just listing the features 😊

Reviewer 3: Proxy search request are important but not a important feature enough to be put here. It would just make the things heavier

After its initial release in v1.3.0, we have now fully stabilized the Vector Store feature. The experimental feature no longer requires manual activation, and the `vectorStore` field is no longer displayed or accepted by the `/experimental-features` route. This makes it even simpler to utilize our hybrid search capability, which delivers significantly better search relevance by combining traditional full text search with AI-powered semantic search.

### Breaking Changes before Stabilization

* Accepted Ollama URLs can only end with `/api/embed` and `/api/embeddings`.
* Modified error codes:
  - `invalid_embedder` has been split into `invalid_search_embedder` and `invalid_similar_embedder`. These codes are returned when the `embedder` parameter of a search (resp. similar) request refers to a non-existing embedder configuration or is not a string.
  - `invalid_hybrid_query` has been renamed to `invalid_search_hybrid_query`. It is returned when the `hybrid` parameter is invalid: contains unknown keys or is not either `null` or an object.

Done by @dureuill in https://github.com/meilisearch/meilisearch/pull/5232 & https://github.com/meilisearch/meilisearch/pull/5234

> ## Dumpless Upgrade: Ease upgrading to the next version without a dump
> 
> TBD by tamo

Reviewer 4: ```suggestion
## Experimental Dumpless Upgrade: Ease upgrading to the next version without a dump

From now on you can upgrade any database in the v1.12 version or more to the latest version without using a dump.
That means it's an in-place, way faster upgrade process that consumes only the minimal amount of RAM or disk required.

To upgrade your v1.12 database to v1.13 runs:
```
./meilisearch --experimental-dumpless-upgrade
```

> [!WARNING]
> Currently, the upgrade can fail, leaving your instance in an unrecoverable state. The Meilisearch team greatly encourages you to make a backup of your database before trying this new feature.

Read more about it [here](https://github.com/orgs/meilisearch/discussions/804).

Done by @irevoire and @dureuill in https://github.com/meilisearch/meilisearch/pull/5264

  * Debug log the channel congestion by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5177
  * Improve unexpected panic message by @irevoire in https://github.com/meilisearch/meilisearch/pull/5300
  * Introduce a compaction subcommand in meilitool by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5235
  * Expose a route to get the file content associated with a task by @Kerollmops in https://github.com/meilisearch/meilisearch/pull/5326

❤️ Thanks again to our external contributors:

> - [Meilisearch](https://github.com/meilisearch/meilisearch): @takaebato, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay, @DerTimonius, @manojks1999, @ellnix, @Guikingone

Reviewer 2: ```suggestion
- [Meilisearch](https://github.com/meilisearch/meilisearch): @takaebato, @Sherlouk, @jameshiew, @K-Kumar-01, @mhmoudr, @eltociear, @Gnosnay, @DerTimonius, @manojks1999, @ellnix, @Guikingone, @migueltarga
```
